File: fsharp/cars-assemble/CarsAssemble.fs
module CarsAssemble

let defaultCarsPerHour = 221

let successRate (speed: int) : float =
    if speed <= 0 then
        0.00
    else if speed >= 1 && speed <= 4 then
        1.00
    else if speed >= 5 && speed <= 8 then
        0.90
    else if speed = 9 then
        0.80
    else
        0.77

let productionRatePerHour (speed: int) : float =
    let prediction = speed * defaultCarsPerHour |> float
    let defectAdjustment = successRate speed
    prediction * defectAdjustment

let workingItemsPerMinute (speed: int) : int =
    productionRatePerHour speed / 60.0 |> int

File: java/cars-assemble/src/main/java/CarsAssemble.java
public class CarsAssemble {
    private static int PRODUCTION_RATE = 221;

    public double productionRatePerHour(int speed) {
        if (speed >= 1 && speed <= 4) {
            return speed * PRODUCTION_RATE;
        } else if (speed >= 5 && speed <= 8) {
            return speed * PRODUCTION_RATE * 0.9;
        } else if (speed == 9) {
            return speed * PRODUCTION_RATE * 0.8;
        } else {
            return speed * PRODUCTION_RATE * 0.77;
        }
    }

    public int workingItemsPerMinute(int speed) {
        return (int) productionRatePerHour(speed) / 60;
    }
}
